File: products/compliance/src/analyzers/ast-analyzer.ts
import type { Analyzer, Finding, RuleDefinition, ScannedFile } from '../types.js';
import { readFileSync } from 'node:fs';

/**
 * Regex-based code pattern analyzer.
 *
 * Scans TypeScript/JavaScript source files for common security anti-patterns
 * such as SQL injection, weak cryptography, empty catch blocks, and
 * exposed stack traces. Each check maps to one or more rule patterns
 * defined in the YAML rule files.
 */

const CODE_EXTENSIONS = new Set(['ts', 'js', 'tsx', 'jsx']);
const MAX_FILE_SIZE = 500 * 1024; // 500 KB

/**
 * A pattern check links a rule-pattern string to one or more regexes
 * that are run against file content (line-by-line or full-content).
 */
interface PatternCheck {
  /** The rule pattern identifier (e.g. 'sql-injection', 'weak-hash'). */
  pattern: string;
  /** Human-readable evidence label. */
  label: string;
  /**
   * Run the check against a file's content and lines.
   * Returns an array of matches with line number and evidence string.
   */
  check(content: string, lines: string[]): PatternMatch[];
}

interface PatternMatch {
  line: number;
  evidence: string;
}

// ── Pattern checks ──────────────────────────────────────────────────────

const PATTERN_CHECKS: PatternCheck[] = [
  // 1. SQL Injection — string concatenation or template-literal interpolation in SQL
  {
    pattern: 'sql-injection',
    label: 'SQL injection via string concatenation',
    check(_content, lines) {
      const matches: PatternMatch[] = [];
      const sqlKeyword = /\b(SELECT|INSERT|UPDATE|DELETE|DROP|ALTER)\b/i;
      for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (!sqlKeyword.test(line)) continue;
        // String concat with +
        if (/['"`].*\b(SELECT|INSERT|UPDATE|DELETE)\b.*['"`]\s*\+/i.test(line) ||
            /\+\s*['"`].*\b(SELECT|INSERT|UPDATE|DELETE|WHERE)\b/i.test(line)) {
          matches.push({ line: i + 1, evidence: 'SQL query built with string concatenation' });
          continue;
        }
        // Template literal with ${...} interpolation inside SQL
        if (/`[^`]*\b(SELECT|INSERT|UPDATE|DELETE)\b[^`]*\$\{[^}]+\}[^`]*`/i.test(line) ||
            /`[^`]*\$\{[^}]+\}[^`]*\b(WHERE|SELECT|INSERT|UPDATE|DELETE)\b[^`]*`/i.test(line)) {
          matches.push({ line: i + 1, evidence: 'SQL query built with template literal interpolation' });
        }
      }
      return matches;
    },
  },

  // 2. Weak Hashing — MD5 or SHA1
  {
    pattern: 'weak-hash',
    label: 'Weak hashing algorithm',
    check(_content, lines) {
      const matches: PatternMatch[] = [];
      const re = /createHash\s*\(\s*['"](?:md5|sha1)['"]\s*\)/i;
      for (let i = 0; i < lines.length; i++) {
        if (re.test(lines[i])) {
          matches.push({ line: i + 1, evidence: `Weak hash: ${lines[i].trim()}` });
        }
      }
      return matches;
    },
  },

  // 3. ECB Mode
  {
    pattern: 'ecb-mode',
    label: 'ECB encryption mode',
    check(_content, lines) {
      const matches: PatternMatch[] = [];
      const re = /createCipheriv\s*\(\s*['"][^'"]*ecb[^'"]*['"]/i;
      for (let i = 0; i < lines.length; i++) {
        if (re.test(lines[i])) {
          matches.push({ line: i + 1, evidence: 'ECB block cipher mode used' });
        }
      }
      return matches;
    },
  },

  // 4. Hardcoded Crypto Key/IV — Buffer.from('...') near createCipheriv in file
  {
    pattern: 'hardcoded-crypto-key',
    label: 'Hardcoded encryption key or IV',
    check(content, lines) {
      const matches: PatternMatch[] = [];
      // Only flag if the file also uses createCipheriv
      if (!/createCipheriv/i.test(content)) return matches;
      const re = /Buffer\.from\s*\(\s*['"][^'"]+['"]\s*\)/;
      for (let i = 0; i < lines.length; i++) {
        if (re.test(lines[i])) {
          matches.push({ line: i + 1, evidence: 'Hardcoded Buffer used with crypto' });
        }
      }
      return matches;
    },
  },

  // 5. Empty Catch Block
  {
    pattern: 'empty-catch',
    label: 'Empty catch block',
    check(content, _lines) {
      const matches: PatternMatch[] = [];
      // Use a regex that matches catch blocks where the body is empty
      // (only whitespace/newlines between { and })
      const re = /catch\s*\([^)]*\)\s*\{[\s]*\}/g;
      let match: RegExpExecArray | null;
      while ((match = re.exec(content)) !== null) {
        // Count line number from content position
        const lineNum = content.slice(0, match.index).split('\n').length;
        matches.push({ line: lineNum, evidence: 'Empty catch block swallows errors' });
      }
      return matches;
    },
  },

  // 6. Stack Trace Exposed — .stack in a response context
  {
    pattern: 'exposed-stack-trace',
    label: 'Stack trace exposed to client',
    check(_content, lines) {
      const matches: PatternMatch[] = [];
      for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (/\.stack\b/.test(line) && /\b(res|response)\b/.test(line)) {
          matches.push({ line: i + 1, evidence: 'Stack trace sent in response' });
        }
      }
      return matches;
    },
  },

  // 7. Sensitive data in logs
  {
    pattern: 'sensitive-in-logs',
    label: 'Sensitive data in logs',
    check(_content, lines) {
      const matches: PatternMatch[] = [];
      const logCall = /\b(console\.(log|info|debug|warn)|logger?\.(log|info|debug|warn))\s*\(/;
      const sensitivePattern = /\b(password|secret|token|apiKey|api_key|ssn|creditCard|credit_card)\b/i;
      for (let i = 0; i < lines.length; i++) {
        if (logCall.test(lines[i]) && sensitivePattern.test(lines[i])) {
          matches.push({ line: i + 1, evidence: 'Sensitive data may be logged' });
        }
      }
      return matches;
    },
  },

  // 8. Console.log in production code
  {
    pattern: 'console-in-production',
    label: 'Console.log in production code',
    check(_content, lines) {
      const matches: PatternMatch[] = [];
      const re = /\bconsole\.(log|info|debug)\s*\(/;
      for (let i = 0; i < lines.length; i++) {
        if (re.test(lines[i])) {
          matches.push({ line: i + 1, evidence: 'console.log found in source' });
        }
      }
      return matches;
    },
  },
];

export class AstAnalyzer implements Analyzer {
  name = 'ast-analyzer';

  async analyze(files: ScannedFile[], rules: RuleDefinition[]): Promise<Finding[]> {
    const myRules = rules.filter((r) => r.analyzer === this.name);
    if (myRules.length === 0) return [];

    // Build lookup from pattern -> rules (multiple rules may share a pattern across frameworks)
    const rulesByPattern = new Map<string, RuleDefinition[]>();
    for (const rule of myRules) {
      const existing = rulesByPattern.get(rule.pattern) ?? [];
      existing.push(rule);
      rulesByPattern.set(rule.pattern, existing);
    }

    const findings: Finding[] = [];

    for (const file of files) {
      // Only scan code files
      if (!CODE_EXTENSIONS.has(file.extension.toLowerCase())) continue;
      // Skip large files
      if (file.size > MAX_FILE_SIZE) continue;

      let content: string;
      try {
        content = readFileSync(file.path, 'utf-8');
      } catch {
        continue;
      }

      // Quick binary check
      if (content.includes('\0')) continue;

      const lines = content.split('\n');

      // Run each pattern check
      for (const check of PATTERN_CHECKS) {
        // Only run if there are rules that care about this pattern
        if (!rulesByPattern.has(check.pattern)) continue;

        const matches = check.check(content, lines);
        for (const match of matches) {
          this.addFindings(findings, check.pattern, rulesByPattern, {
            file: file.relativePath,
            line: match.line,
            evidence: match.evidence,
          });
        }
      }
    }

    return findings;
  }

  /**
   * Create findings for all rules matching a given pattern.
   * A single pattern may map to rules in multiple frameworks.
   */
  private addFindings(
    findings: Finding[],
    pattern: string,
    rulesByPattern: Map<string, RuleDefinition[]>,
    context: { file?: string; line?: number; evidence?: string },
  ): void {
    const matchedRules = rulesByPattern.get(pattern);
    if (!matchedRules || matchedRules.length === 0) return;

    for (const rule of matchedRules) {
      findings.push({
        id: rule.id,
        title: rule.title,
        description: rule.description,
        severity: rule.severity,
        framework: rule.framework,
        controlIds: rule.controls,
        file: context.file,
        line: context.line,
        evidence: context.evidence,
        analyzer: this.name,
        fixable: rule.fixable,
      });
    }
  }
}

File: products/compliance/analyzers/ast_analyzer.go
package analyzers

import (
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/rishav1305/soul/products/compliance/rules"
)

// ASTAnalyzer implements the Analyzer interface for regex-based source code
// anti-pattern detection. Despite the name, it uses regex patterns rather
// than actual AST parsing.
type ASTAnalyzer struct{}

// Name returns the analyzer identifier used to match rules.
func (a *ASTAnalyzer) Name() string {
	return "ast-analyzer"
}

// astPattern links a compiled regex to a rule pattern and descriptive name.
type astPattern struct {
	name        string
	regex       *regexp.Regexp
	rulePattern string
	// skipFunc optionally filters out false positives on matched lines.
	skipFunc func(line string, match string) bool
}

// astPatterns contains all regex patterns for anti-pattern detection.
// Note: these patterns detect security anti-patterns in scanned codebases.
var astPatterns = []astPattern{
	// 1. Detects dangerous eval() calls in scanned code
	{
		name:        "eval() usage",
		regex:       regexp.MustCompile(`\beval\s*\(`),
		rulePattern: "eval-usage",
	},
	// 2. SQL string concatenation
	{
		name:        "SQL injection risk",
		regex:       regexp.MustCompile(`(?i)(?:query.*\+.*req\.|SELECT.*\+)`),
		rulePattern: "sql-injection",
	},
	// 3. Weak crypto (MD5/SHA1)
	{
		name:        "Weak crypto hash",
		regex:       regexp.MustCompile(`createHash\s*\(\s*['"](?:md5|sha1)['"]\s*\)`),
		rulePattern: "weak-hash",
	},
	// 4. Insecure random (Math.random())
	{
		name:        "Insecure random",
		regex:       regexp.MustCompile(`Math\.random\(\)`),
		rulePattern: "insecure-random",
	},
	// 5. XSS risk (innerHTML / dangerouslySetInnerHTML)
	{
		name:        "XSS risk",
		regex:       regexp.MustCompile(`(?:innerHTML|dangerouslySetInnerHTML)`),
		rulePattern: "xss-risk",
	},
	// 6. Disabled SSL verification
	{
		name:        "SSL verification disabled",
		regex:       regexp.MustCompile(`rejectUnauthorized\s*:\s*false`),
		rulePattern: "ssl-disabled",
	},
	// 7. Hardcoded IP addresses (skip 0.0.0.0 and 127.0.0.1)
	{
		name:        "Hardcoded IP address",
		regex:       regexp.MustCompile(`\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b`),
		rulePattern: "hardcoded-ip",
		skipFunc: func(line string, match string) bool {
			return match == "0.0.0.0" || match == "127.0.0.1"
		},
	},
	// 8. Empty catch blocks
	{
		name:        "Empty catch block",
		regex:       regexp.MustCompile(`catch\s*\([^)]*\)\s*\{\s*\}`),
		rulePattern: "empty-catch",
	},
}

// astSourceExtensions lists file extensions the AST analyzer inspects.
var astSourceExtensions = map[string]bool{
	"ts":   true,
	"js":   true,
	"py":   true,
	"go":   true,
	"java": true,
	"rb":   true,
}

// Analyze scans source files line by line for anti-patterns.
func (a *ASTAnalyzer) Analyze(files []ScannedFile, allRules []rules.Rule) ([]Finding, error) {
	myRules := filterRules(allRules, a.Name())
	if len(myRules) == 0 {
		return nil, nil
	}

	rulesByPattern := groupRulesByPattern(myRules)
	var findings []Finding

	for _, file := range files {
		if !astSourceExtensions[strings.ToLower(file.Extension)] {
			continue
		}
		if file.Size > maxFileSize {
			continue
		}

		content, err := os.ReadFile(file.Path)
		if err != nil {
			continue
		}

		if containsNullByte(content) {
			continue
		}

		lines := strings.Split(string(content), "\n")
		for lineIdx, line := range lines {
			for _, pat := range astPatterns {
				matches := pat.regex.FindAllString(line, -1)
				for _, m := range matches {
					// Apply skip function if defined
					if pat.skipFunc != nil && pat.skipFunc(line, m) {
						continue
					}

					matchedRules, ok := rulesByPattern[pat.rulePattern]
					if !ok || len(matchedRules) == 0 {
						continue
					}

					for _, rule := range matchedRules {
						findings = append(findings, Finding{
							ID:          rule.ID,
							Title:       rule.Title,
							Description: fmt.Sprintf("%s (%s)", rule.Description, pat.name),
							Severity:    rule.Severity,
							Framework:   rule.Framework,
							ControlIDs:  rule.Controls,
							File:        file.RelativePath,
							Line:        lineIdx + 1,
							Column:      strings.Index(line, m) + 1,
							Evidence:    strings.TrimSpace(line),
							Analyzer:    a.Name(),
							Fixable:     rule.Fixable,
						})
					}
				}
			}
		}
	}

	return findings, nil
}
